File: LeetCode/iter1/c++/705.design-hash-set.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <queue>
#include <stack>

using namespace std;

const int SIZE = 1000000;

class MyHashSet {
public:
	/** Initialize your data structure here. */
	MyHashSet() {
		arr = new int[SIZE];
		for (int i = 0; i < SIZE; i++) {
			arr[i] = 0;
		}
	}

	void add(int key) {
		arr[key] = 1;
	}

	void remove(int key) {
		arr[key] = 0;
	}

	/** Returns true if this set contains the specified element */
	bool contains(int key) {
		return arr[key] == 1;
	}

private:
	int* arr;
};

int main()
{
	MyHashSet myHashSet;

	myHashSet.add(1);
	myHashSet.add(2);
	cout << myHashSet.contains(1) << endl;    // returns true
	cout << myHashSet.contains(3) << endl;    // returns false (not found)
	myHashSet.add(2);
	cout << myHashSet.contains(2) << endl;    // returns true
	myHashSet.remove(2);
	cout << myHashSet.contains(2) << endl;    // returns false (already removed)

	return 0;
}

File: LeetCode/iter1/java/705.design-hash-set.java
package org.ironman.base.playground.leetcode;

public class Solution {

    public static void main(String[] args) {
        MyHashSet hashSet = new MyHashSet();
        hashSet.add(1);
        hashSet.add(2);
        System.out.println(hashSet.contains(1));    // returns true
        System.out.println(hashSet.contains(3));    // returns false (not found)
        hashSet.add(2);
        System.out.println(hashSet.contains(2));    // returns true
        hashSet.remove(2);
        System.out.println(hashSet.contains(2));    // returns false (already removed)
    }

    static class MyHashSet {

        private int[] arr;
        private static final int SIZE = 1000000;
        private static final int USED = 1;
        private static final int NOT_FOUND = 0;
        /**
         * Initialize your data structure here.
         */
        public MyHashSet() {
            arr = new int[SIZE];
            for (int i = 0; i < SIZE; i++) {
                arr[i] = NOT_FOUND;
            }
        }

        public void add(int key) {
            arr[key] = USED;
        }

        public void remove(int key) {
            arr[key] = NOT_FOUND;
        }

        /**
         * Returns true if this set contains the specified element
         */
        public boolean contains(int key) {
            return arr[key] == USED;
        }
    }

}